python/cuspatial/cuspatial/core/spatial/distance.py

```diff
@@ -124,6 +126,87 @@ def haversine_distance(p1_lon, p1_lat, p2_lon, p2_lat):
     return cpp_haversine_distance(p1_lon, p1_lat, p2_lon, p2_lat)
 
 
+def pairwise_point_distance(points1: GeoSeries, points2: GeoSeries):
+    """Compute shortest distance between pairs of points and multipoints
+
+    Parameters
+    ----------
+    points1 : GeoSeries
+        lhs of points or multipoints
+    points2 : GeoSeries
+        A GeoSeries of points or multipoints
+
+    Returns
+    -------
+    distance : cudf.Series
+        the distance between each pair of points
+
+    Examples
+    --------
+    >>> import geopandas as gpd
+    >>> cities = gpd.read_file(gpd.datasets.get_path('naturalearth_cities'))
+    >>> cities
+                name                     geometry
+    0    Vatican City    POINT (12.45339 41.90328)
+    1      San Marino    POINT (12.44177 43.93610)
+    2           Vaduz     POINT (9.51667 47.13372)
+    3      Luxembourg     POINT (6.13000 49.61166)
+    4         Palikir    POINT (158.14997 6.91664)
+    ..            ...                          ...
+    197         Cairo    POINT (31.24802 30.05191)
+    198         Tokyo   POINT (139.74946 35.68696)
+    199         Paris     POINT (2.33139 48.86864)
+    200      Santiago  POINT (-70.66899 -33.44807)
+    201     Singapore    POINT (103.85387 1.29498)
+
+    [202 rows x 2 columns]
+    >>> cities0 = cuspatial.from_geopandas(cities.geometry[0:10])
+    >>> cities1 = cuspatial.from_geopandas(cities.geometry[10:20])
+    >>> cuspatial.pairwise_point_distance(cities0, cities1)
+    0     61.818996
+    1     11.019265
+    2     79.854261
+    3     56.844978
+    4    131.292340
+    5    139.818452
+    6    173.507174
+    7    124.346079
+    8     47.829125
+    9    159.156985
+    dtype: float64
+    """
```

Comment:
This is not a good example particularly because I don't think that cartesian distance in the haversine coordinate space is especially meaningful. You could use `lonlat_to_cartesian` first in order to give the points meter values, as one suggestion.

Reply:
Yeah but that also doesn't make sense right? `lonlat_to_cartesian` (aka sinusoidal projection) assumes all points are the in proximity of the intersection of equator and prime meridian. The points we get from the cities are all over places. Maybe I should instead use a handcrafted example here?

Reply:
That's true, lonlat_to_cartesian across the entire globe would produce nonsense. Maybe just a set of points on a line pairwised against a set of points on a similar line. Something easy would be like [0, 0], [1, 1], [2, 2] ... in one direction and [0, 0], [-1, 1], [-2, 2], ... which would be easy to understand and visualize.